python/lsst/meas/algorithms/brightStarStamps.py

```diff
@@ -0,0 +1,280 @@
+        recovered from the metadata.
+
+        Returns
+        -------
+        annularFluxes : list[`float`]
+        """
+        return [stamp["annularFlux"] for stamp in self._starStamps]
+
+    def selectByMag(self, magMin=None, magMax=None):
+        """Return the subset of bright star stamps for objects with specified
+        magnitude cuts (in Gaia G).
+
+        Parameters
+        ----------
+        magMin : `float`, optional
+            Keep only stars fainter than this value.
+        magMax : `float`, optional
+            Keep only stars brighter than this value.
+        """
+        selector = np.array([True for _ in range(len(self))])
+        if magMin is not None:
+            selector *= np.array(self.getMagnitudes()) > magMin
+        if magMax is not None:
+            selector *= np.array(self.getMagnitudes()) < magMax
+        selIdx = np.where(selector)[0]
+        return BrightStarStamps([self._starStamps[j] for j in selIdx], metadata=self._metadata)
+
+    def getMetadata(self):
+        """Retrieve metadata associated with these `Defects`.
+
+        Returns
+        -------
+        meta : `lsst.daf.base.PropertyList`
+            Metadata. The returned `~lsst.daf.base.PropertyList` can be
+            modified by the caller and the changes will be written to
+            external files.
+        """
+        return self._metadata
+
+    def _setMetadata(self, metadata=None):
+        if metadata is None:
+            self._metadata = PropertySet()
+        else:
+            self._metadata = metadata
+
+    def _checkStarStamp(self, stamp):
+        """Ensures the provided star stamp contains the necessary keys.
+        """
+        necKeys = {"starStamp", "gaiaGMag", "gaiaId", "annularFlux"}
+        if not necKeys <= stamp.keys():
+            raise ValueError(f"The provided starStamp is missing necessary keys ({necKeys})")
+        return stamp
+
+    def _setAnnulusRadii(self, innerRadius, outerRadius):
+        self._checkRadius(innerRadius, "INNER_RADIUS")
+        self._checkRadius(outerRadius, "OUTER_RADIUS")
+        return None
+
+    def _checkRadius(self, radiusValue, metadataName):
+        """Ensure provided annulus radius is consistent with that present
+        in metadata. If metadata does not contain annulus radius, add it.
+        """
+        # if a radius value is already present in metadata, ensure it matches
+        # the one given
+        if self._metadata.exists(metadataName):
+            if radiusValue is not None:
+                if self._metadata.get(name=metadataName) != radiusValue:
+                    raise AttributeError("BrightStarStamps instance already contains different annulus radii "
+                                         + f"values ({metadataName}).")
+        # if not already in metadata, a value must be provided
+        elif radiusValue is None:
+            raise AttributeError("No radius value provided for the AnnularFlux measurement "
```

Comment:
This msg should say something about no value provided, and none present in the metadata, so the user can track down that there are two different places it could come from

Reply:
To be honest I am not entirely sure why this exists, the user must already supply the meta data field, and the calculations for the stamps must be done prior to creating this class (I guess append could add some if you had an empty list) The radii should be known at the point of creating this class then. Why have the arguments instead of just enforcing they are in the meta data passed in. Is it just a convenience for the user? I think the class would be simpler if the radii just came in with the meta data. Its a bit confusing to have different ways of setting it, the user might not know which one to use, or think they must use both.

Reply:
Right, this is confusing as I tried to take into account how these objects will be generated and used. There are a lot of disparate elements that led me to write things this way:
* These `starStamps` objects have already been normalized, using their annularFlux value, for set values of annular radii. It only makes sense to store them together within one `BrightStarStamps` instance if they have all used the same annulus definition.
* We will want to write a great many of these to disk; we need to include the radii used in the written fits. It seems the best way to do so is to add these values to the metadata, that will then be set within the `writeFits` method (line 233 in this version of the file).
* As discussed in the thread about the metadata getter, `self._metadata` is likely to be the full metadata of the `Exposure` from which the stamps were extracted, with extra fields added here. The “user” who creates these instances (DM-25304) has this readily available on the one hand, and the radii values he used on the other. I thought it would make sense to have the class itself merge the two internally, so this user only needs to dump the exposure metadata, and the two radii values, at initialization.
* Another use case will be the reading, concatenation, and further use (DM-25305) of several BrightStarStamps, saved to disk. We very much want to ensure we are concatenating bright star stamps that used the same annular flux radii; but since we’re reading them from disk, that information is in the metadata now.

I do think the duplication here is good, or at least necessary: we want users to be able to provide their annular radii when they create a new instance, and dump metadata at the same time. If that metadata comes from another BrightStarStamps (likely read from disk), right then is a good time to fail if the radii used do not match those in the metadata. Otherwise, they can safely be added.

Having written all this, however, I realize this would perhaps look a little less out-of-place with an `extend` method, so I added one. This is what will be used for concatenation in DM-25304, rather than `append`; and it makes it clearer (I think) why we want this `_checkRadius` there. This led me to shuffling things around a little; `_checkRadius` is called directly in the init now, and the two radii are also kept as private members.

Does that make sense?

Reply:
Yes it does. Its really just a question of who is responsible for adding these values to the metadata. If you want the class to handle it in the case of arguments as a convenience to the user, that's fine. I do like the addition of the extend method. However, in regards to your first built point, you actually have no mechanism in place to verify that the radii are the same in the append method. You should either require radii as args, put a big warning in the method docstring, or require adding elements be done by calling extend with a new instance.